io-engine/src/bdev/nexus/nexus_channel.rs

```diff
@@ -193,67 +210,70 @@ impl<'n> NexusChannel<'n> {
         debug!("{self:?}: device '{device_name}' disconnected");
     }
 
-    /// Refreshing our channels simply means that we either have a child going
-    /// online or offline. We don't know which child has gone, or was added, so
-    /// we simply put back all the channels, and reopen the bdevs that are in
-    /// the online state.
-    pub(crate) fn reconnect_all(&mut self) {
-        // clear the vector of channels and reset other internal values,
-        // clearing the values will drop any existing handles in the
-        // channel
+    /// We receive this callback as a result of some reconfigure event regarding
+    /// a child. Depending on the event type and the child state, we'd
+    /// adjust the impacted child as a valid reader or writer in the channel.
+    /// ## ChildRebuildBegin -
+    /// newly added child or child undergoing partial rebuild.
+    /// ## ChildRebuildEnd -
+    /// a rebuild finished, may be successful or failed/stopped.
+    /// ## ChildUnplug -
+    /// a child device is being unplugged and destroyed.
+    pub(crate) fn reconfigure_channel(&mut self, dev_ctx: &DrEvent) {
         self.previous_reader = UnsafeCell::new(0);
 
-        // nvmx will drop the I/O qpairs which is different from all other
-        // bdevs we might be dealing with. So instead of clearing and refreshing
-        // which had no side effects before, we create a new vector and
-        // swap them out later
-
-        let mut writers = Vec::new();
-        let mut readers = Vec::new();
-
-        // iterate over all our children which are in the healthy state
-        self.nexus()
-            .children_iter()
-            .filter(|c| c.is_healthy())
-            .for_each(|c| match (c.get_io_handle(), c.get_io_handle()) {
-                (Ok(w), Ok(r)) => {
-                    writers.push(w);
-                    readers.push(r);
-                }
-                _ => {
-                    c.set_faulted_state(FaultReason::CantOpen);
-                    error!("{self:?}: failed to get I/O handle for {c:?}");
-                }
-            });
-
-        // then add write-only children
-        if !readers.is_empty() {
-            self.nexus()
-                .children_iter()
-                .filter(|c| c.is_rebuilding())
-                .for_each(|c| match c.get_io_handle() {
-                    Ok(hdl) => {
-                        debug!(
-                            "{self:?}: connecting child device \
-                                in write-only mode: {c:?}"
-                        );
-                        writers.push(hdl);
+        match dev_ctx {
+            DrEvent::ChildRebuildBegin(dev) => {
+                if let Ok(child) = self.nexus().child_by_device(dev.as_str()) {
+                    debug_assert!(!self.is_writer(dev.as_str()));
+                    if !self.readers.is_empty() && !self.is_writer(dev.as_str())
```

Comment:
"!self.is_writer(dev.as_str()" is always true here, why adding it as a condition?